mvpa2/algorithms/searchlight_hyperalignment.py

```diff
@@ -0,0 +1,476 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Searchlight-based hyperalignment"""
+import numpy as np
+from mvpa2.base.state import ClassWithCollections
+from mvpa2.base.param import Parameter
+from mvpa2.algorithms.hyperalignment import Hyperalignment
+from mvpa2.measures.base import Measure
+from mvpa2.datasets import Dataset, vstack
+from mvpa2.mappers.staticprojection import StaticProjectionMapper
+from scipy.linalg import LinAlgError
+from mvpa2.measures.searchlight import sphere_searchlight
+from mvpa2.mappers.zscore import zscore
+from scipy.sparse import coo_matrix, dok_matrix, csc_matrix
+from mvpa2.featsel.helpers import FractionTailSelector, FixedNElementTailSelector
+
+if __debug__:
+    from mvpa2.base import debug
+
+def compute_feature_scores(datasets, exclude_from_model=None):
+    """
+    Takes a list of datasets and computes a magical feature
+    score for each feature in each dataset
+    :ref:`Haxby et al., Neuron (2011) <HGC+11>`
+    Parameters
+    ----------
+        datasets : list or tuple of datasets
+        
+        exclude_from_model: list of dataset indices that won't participate
+                    in voxel selection of others
+
+    Returns
+    -------
+        list : a list of feature scores; higher the score, better the feature
+
+    NOTE: This function assumes that the datasets are zscored
+    """
+    if exclude_from_model is None:
+        exclude_from_model=[]
+    feature_scores = [ np.zeros(sd.nfeatures) for sd in datasets ]
+    for i, sd in enumerate(datasets):
+        for j, sd2 in enumerate(datasets[i+1:]):
+            corr_temp= np.dot(sd.samples.T, sd2.samples)
+            if  j+i+1 not in exclude_from_model:                
+                feature_scores[i] = feature_scores[i] + np.max(corr_temp, axis = 1)
+            if i not in exclude_from_model:
+                feature_scores[j+i+1] = feature_scores[j+i+1] + np.max(corr_temp, axis = 0)
```

Comment:
in both of above, why not just to use `+=` operation?  should be more eficient and produce more concise thus more readable code